craft_parts/plugins/maven_plugin.py

```diff
@@ -74,6 +77,97 @@ def validate_environment(
                 reason=f"invalid maven version {version!r}",
             )
 
+        system_java_major_version = self._get_system_openjdk_java_major_version()
+        project_java_major_version = self._get_project_java_major_version()
+        # if the project Java version cannot be detected, let it try build anyway.
+        if project_java_major_version is None:
+            return
+
+        if system_java_major_version < project_java_major_version:
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason="system Java version is less than project Java version."
+                f"system: {system_java_major_version}, project: {project_java_major_version}",
+            )
+
+    def _get_project_java_major_version(self) -> int | None:
+        options = cast(MavenPluginProperties, self._options)
+        maven_executable = "./mvnw" if options.maven_use_mvnw else "mvn"
+        effective_pom_path = Path("./effective.pom")
+        try:
+            self._execute(
+                # Write to file since writing to stdout without maven info logs require sudo, i.e.
+                # sudo mvn help:effective-pom -q -DforceStdout -doutput=/dev/stdout
+                f"{maven_executable} help:effective-pom -Doutput={effective_pom_path}"
+            )
+            project_java_major_version = _parse_project_java_version(effective_pom_path)
+            return _extract_java_version(project_java_major_version)
+        except subprocess.CalledProcessError as err:
```

Comment:
Hmm, to me it is a bit confusing how `CalledProcessError` could be raised here since it isn't directly in the code here. I guess that it could come from the `_execute` method? In general, I would rewrap the exception though I'm not sure what the style is in this project